source/agora/node/admin/QRCodeInterface.d

```diff
@@ -0,0 +1,269 @@
+    {
+        this.key_pair = key_pair;
+    }
+
+    /***************************************************************************
+
+        Respond with QR code containing login information.
+
+        Enter the name of the APP where the login information will be
+        used as the get query.
+
+        Convert LoginInfo to JSON format and respond with the QR code
+        of SVG format.
+        All public key cryptography is done on Curve25519.
+        The signature of the vectorCard provides verification function in the SDK.
+        This LoginInfo has an expiration time.
+        It provide a randomly generated temporary private key encoded
+        in strings on demand.
+        In the future, there is a need to enter a temporary key.
+
+    ***************************************************************************/
+
+    public void loginQRCreate (scope HTTPServerRequest req,
+        scope HTTPServerResponse res)
+    {
+        // The randomly generated temporary KeyPair
+        KeyPair temp_kp = KeyPair.random();
+
+        SysTime expires_time = Clock.currTime();
+        // A 'nonce' used to allow expiring
+        expires_time += 90.days;
+
+        LoginInfo login_info = LoginInfo(
+            temp_kp.secret.toString(PrintMode.Clear),
+            VoterCard(
+                this.key_pair.address,
+                temp_kp.address,
+                expires_time.toISOExtString()
+            )
+        );
+
+        login_info.voter_card.signature =
+            this.key_pair.secret.sign(hashFull(login_info)[]);
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(login_info));
+    }
+
+    /***************************************************************************
+
+        Respond with QR code containing EncryptionKey.
+
+        Enter the name of the APP & height where the encryptionKey will be used
+        as the get query.
+
+        The encryption key is hashed using the blake2 cryptographic
+        hash function.
+        The encryption key is derived from the pre-image as follows:
+        Hash(pre_image, "APPNAME");
+
+    ***************************************************************************/
+
+    public void keyQRCreate (scope HTTPServerRequest req,
+        scope HTTPServerResponse res)
+    {
+        enforceHTTP("app" in req.query && "height" in req.query,
+            HTTPStatus.badRequest, "Missing app/height field.");
+
+        auto app = req.query["app"];
+        auto height = Height.fromString(req.query["height"]);
+
+        // TODO: PreImage to be released at that height
+        // Or it can be pre-released, so it can be `height + PreimageRevealPeriod`
+        PreImageInfo pre_image;
+
+        // blake2b hashing
+        const Hash key = hashMulti(pre_image, app);
+
+        EncryptionKey encryptionKey = EncryptionKey(
+            app,
+            height,
+            key
+        );
+
```

Comment:
The protocol ID must be entered as get parameter.
Hash key and proposal ID once again. The final value is the encryption key.
https://github.com/bpfkorea/infrastructure/tree/master/docs/votera#getting-an-encryption-key

Reply:
It has to be hashed at the Votera app.
`Hash prop_encryption_key = blake2bHash(encryption_key, propID);`